pkg/piping/sync_async_piping_manager.go

```diff
@@ -0,0 +1,172 @@
+package piping
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/regionless-storage-service/pkg/config"
+	"github.com/regionless-storage-service/pkg/constants"
+	"github.com/regionless-storage-service/pkg/database"
+	"github.com/regionless-storage-service/pkg/index"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+)
+
+type SyncAsyncPiping struct {
+	databaseType constants.StoreType
+}
+
+func NewSyncAsyncPiping(storeType constants.StoreType) *SyncAsyncPiping {
+	return &SyncAsyncPiping{databaseType: storeType}
+}
+
+func (sap *SyncAsyncPiping) Read(ctx context.Context, rev index.Revision) (string, error) {
+	_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "SyncAsyncPiping Read")
+	defer rootSpan.End()
+	syncNodes, asyncNodes, err := splitStores(rev.GetNodes())
+	if err != nil {
+		return "", err
+	}
+	target := ""
+	if len(syncNodes) > 0 {
+		target = syncNodes[0]
+	} else if len(asyncNodes) > 0 {
+		target = asyncNodes[0]
+	} else {
+		return "", fmt.Errorf("the rev %v does not have any nodes", rev)
+	}
+	// The first sync store has the fewest latency. Threfore, it is chosen to read
+	if database, err := database.FactoryWithNameAndLatency(sap.databaseType, target, 0); err != nil {
+		return "", err
+	} else {
+		return database.Get(rev.String())
+	}
+}
+
+func (sap *SyncAsyncPiping) Write(ctx context.Context, rev index.Revision, val string) error {
+	_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "SyncAsyncPiping write")
+	defer rootSpan.End()
+	syncNodes, asyncNodes, err := splitStores(rev.GetNodes())
+	if err != nil {
+		return err
+	}
+
+	for _, asyncNode := range asyncNodes {
+		if len(asyncNode) < 1 {
+			continue
+		}
+		go func(ctx context.Context, databaseType constants.StoreType, name, key, val string) {
+			_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "async db put")
+			defer rootSpan.End()
+			if database, err := database.FactoryWithNameAndLatency(sap.databaseType, name, 0); err != nil {
```

Comment:
is it possible to avoid calling factory method which typically seems creating new object? this is a database lookup map at database.Storages; maybe we can leverage it.

Reply:
I prefer using Factory to get different instances based on their types. 

For redis, it uses pooling which might be different from others. It might be better to refactor codes once it becomes stab le. Now we are still add new features.